controllers/novacell_controller.go

```
@@ -319,15 +325,59 @@ func (r *NovaCellReconciler) ensureConductor(
 	return ctrl.Result{}, nil
 }
 
+func getNoVNCProxyName(instance *novav1.NovaCell) types.NamespacedName {
+	return types.NamespacedName{Namespace: instance.Namespace, Name: instance.Name + "-novncproxy"}
+}
+
 func (r *NovaCellReconciler) ensureNoVNCProxy(
 	ctx context.Context,
 	h *helper.Helper,
 	instance *novav1.NovaCell,
 ) (ctrl.Result, error) {
+	// There is a case when the user manually created a NoVNCProxy while it
+	// was disabled in the cell and then tries to enable it in the cell.
+	// One can think that this means the cell adopts the NoVNCProxy and
+	// starts managing it.
+	// However at least the label selector of the StatefulSet spec is
+	// immutable, so the cell cannot simply start adopting an existing
+	// NoVNCProxy. But instead the Cell CR will be in error state until the
+	// human deletes the manually created NoVNCProxy and the the Cell will
+	// create its own NoVNCProxy.
+	// See https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates
+
+	novncproxyName := getNoVNCProxyName(instance)
+	novncproxy := &novav1.NovaNoVNCProxy{}
+	err := r.Client.Get(ctx, novncproxyName, novncproxy)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	// If it is not created by us, we don't touch it
+	if !k8s_errors.IsNotFound(err) && !OwnedBy(novncproxy, instance) {
```

Just considering out loud: Would it be a good idea to perform this type of validation within the webhook to prevent entering an 'error' state?

---

Do you mean rejecting a NovaNoVNCProxy CR creation if it is not having an owner reference set to a NovaCell CR instance? We could do that in the validation webhook but I think we don't want to prevent a standalone NovaNoVNCProxy CR creation by a human user (or the envtest test case).

---

Yes I was thinking about rejecting NovaNoVNCProxy CR with wrong ownership, but maybe when we set in NovaCell enable = true we can check in update webhook if already is novncproxy without NovaCell ownership and reject NovaCell update with NovaNoVNCProxy template with enable = true. In that case we will 'fail fast' and user will get error earlier. But I have doubts about whether this check within the competency of the webhook in our case.

---

Ahh I see your point. So basically detecting the manually created NovaNoVNCProxy at NovaCell create or update. This would require that the webhook logic try to reads stuff from the kube-apiserver to see if NovaNoVNCProxy exists with the same name the NovaCell would like to use. 

The doc https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#best-practices-and-warnings does not implicitly forbid  doing such thing in a validating webhook. But:
* from that doc: `It is recommended that admission webhooks should evaluate as quickly as possible (typically in milliseconds)` Reading from the kube-apiserver would take significant time compared the to current processing of the webhook.
* in the context of the webhook code we don't have an api client set up with credentials to talk to the kube-apiserver
* even if the validation webhook would see that there is no NovaNoVNCProxy CR with the name in question we cannot assume in the Reconciler that such CR is not created in the meantime between the NovaCell validation webhook run and the NovaCell Reconcile call. So the logic in the Reconciler still need to be kept. 

At the moment I think it is too much additional complexity for a limited additional value.

---

Agree